xls/contrib/integrator/ir_integrator.cc

```diff
@@ -18,6 +18,122 @@
+  std::string function_name = std::string("IntegrationFunction") +
+                              std::to_string(integration_functions_count++);
+  integration_function->function_ =
+      absl::make_unique<Function>(function_name, package);
+
+  // Package source function parameters as tuple parameters to integration
+  // function.
+  for (const auto* source_func : source_functions) {
+    // Add tuple paramter for source function.
+    std::vector<Type*> arg_types;
+    for (const Node* param : source_func->params()) {
+      arg_types.push_back(param->GetType());
+    }
+    Type* args_tuple_type = package->GetTupleType(arg_types);
+    std::string tuple_name = source_func->name() + std::string("ParamTuple");
+    XLS_ASSIGN_OR_RETURN(
+        Node * args_tuple,
+        integration_function->function_->MakeNodeWithName<Param>(
+            /*loc=*/std::nullopt, tuple_name, args_tuple_type));
+
+    // Add TupleIndex nodes inside function to unpack tuple parameter.
+    int64 paramter_index = 0;
+    for (const Node* param : source_func->params()) {
+      XLS_ASSIGN_OR_RETURN(
+          Node * tuple_index,
+          integration_function->function_->MakeNode<TupleIndex>(
+              /*loc=*/std::nullopt, args_tuple, paramter_index));
+      XLS_RETURN_IF_ERROR(
+          integration_function->SetNodeMapping(param, tuple_index));
+      paramter_index++;
+    }
+  }
+
+  return std::move(integration_function);
+}
+
+absl::Status IntegrationFunction::SetNodeMapping(const Node* source,
+                                                 Node* map_target) {
+  // Validate map pairing.
+  if (source == map_target) {
+    return absl::InternalError("Tried to map a node to itself");
+  }
+  if (!IntegrationFunctionOwnsNode(map_target)) {
+    return absl::InternalError(
+        "Tried to map to a node not owned by the integration function");
+  }
+  // TODO(jbaileyhandle): Reasonable assumption for short-term use cases.  May
+  // be worth relaxing to enable some optimizations of some sort?
+  if (IntegrationFunctionOwnsNode(source) && !IsMappingTarget(source)) {
+    return absl::InternalError(
+        "Tried to map an integration function node that is not itself a "
+        "mapping target");
+  }
+
+  // 'original' is itself a member of the integrated function.
+  if (IntegrationFunctionOwnsNode(source)) {
+    absl::flat_hash_set<const Node*>& nodes_that_map_to_source =
+        integrated_node_to_original_nodes_map_.at(source);
+
+    // Nodes that previously mapped to original now map to map_target.
+    for (const Node* original_node : nodes_that_map_to_source) {
+      integrated_node_to_original_nodes_map_[map_target].insert(original_node);
+      XLS_RET_CHECK(HasMapping(original_node));
+      original_node_to_integrated_node_map_[original_node] = map_target;
+    }
+
+    // No nodes map to source anymore.
+    integrated_node_to_original_nodes_map_.erase(source);
+
+    // 'source' is an external node.
```

Comment:
Does this comment belong in the else block?

Reply:
Describes the else block.  Should be shifted left to match comment for if block.